internal/operator/internal/operator_installer.go

```diff
@@ -62,3 +66,67 @@ func (o OperatorInstaller) InstallOperator(ctx context.Context) (*v1alpha1.Clust
 
 	return todo, nil
 }
+
+func (o OperatorInstaller) verifyOperator(ctx context.Context, ip *v1alpha1.InstallPlan) (v1alpha1.ClusterServiceVersionPhase, error) {
+	ipKey := types.NamespacedName{
+		Namespace: ip.GetNamespace(),
+		Name:      ip.GetName(),
+	}
+	if err := wait.PollImmediateUntil(time.Millisecond*250, func() (bool, error) {
+		if err := o.cfg.Client.Get(ctx, ipKey, ip); err != nil {
+			return false, err
+		}
+		if ip.Status.Phase == v1alpha1.InstallPlanPhaseComplete {
+			return true, nil
+		}
+		return false, nil
+	}, ctx.Done()); err != nil {
+		return v1alpha1.InstallPlanFailed, fmt.Errorf("waiting for operator installation to complete: %v", err)
+	}
+
+	csvKey := types.NamespacedName{
+		Namespace: o.cfg.Namespace,
+	}
+
+	if csvKey.Name == "" {
+		return v1alpha1.InstallPlanFailed, fmt.Errorf("could not find installed CSV in install plan")
+	}
+
+	// verify CSV
+	csvPhase, err := o.verifyCSV(ctx, csvKey)
+	if err != nil {
+		return v1alpha1.CSVPhaseFailed, fmt.Errorf("waiting for CSV installation to complete: %v", err)
+	}
+	return csvPhase, nil
+}
+
+func (o OperatorInstaller) verifyCSV(ctx context.Context, csvKey types.NamespacedName) (v1alpha1.ClusterServiceVersionPhase, error) {
+
+	csv := &v1alpha1.ClusterServiceVersion{}
+	if err := wait.PollImmediateUntil(time.Millisecond*250, func() (bool, error) {
+		if err := o.cfg.Client.Get(ctx, csvKey, csv); err != nil {
+			return false, fmt.Errorf("get clusterserviceversion: %v", err)
+		}
+		if csv.Status.Phase == v1alpha1.CSVPhaseSucceeded {
+			return true, nil
+		}
+		return false, nil
+	}, ctx.Done()); err != nil {
```

Comment:
I'm confused here, when failed to get the CSV, why still need to check the Deployment? Based on my understanding, both CSV and deployment work well, we can say the operator is installed successfully. Anyone of them failed, we can say the operator failed to install.

Reply:
As I understand, this is to get the deployment reason for failure.
@joelanford Is that what you were suggesting?

Reply:
The idea is that if the CSV never succeeds, we should try to be helpful and tell a user _why_ the CSV didn't succeed.

One common reason is that the operator deployment hasn't rolled out.

I think we should only check the deployment status if the error from the `wait.PollImmediateUntil` call is `context.DeadlineExceeded`.

Reply:
Thank you! I see now, yeah, we should try to return the helpful info. Besides the Deployment object, the rule policy(role, clusterrole, rolebinding ...) leads to many failures.